File: app/src/main/java/com/spidroid/starry/ui/home/FeedViewModel.java
package com.spidroid.starry.ui.home;

import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.ViewModel;
import com.google.firebase.firestore.FirebaseFirestore;
import com.spidroid.starry.models.PostModel;
import java.util.ArrayList;
import java.util.List;

public class FeedViewModel extends ViewModel {
  private final MutableLiveData<List<PostModel>> posts = new MutableLiveData<>();
  private final MutableLiveData<String> errors = new MutableLiveData<>();
  private boolean hasLoaded = false;

  public boolean hasPosts() {
    return posts.getValue() != null && !posts.getValue().isEmpty();
  }

  public void setPosts(List<PostModel> newPosts) {
    hasLoaded = true;
    posts.setValue(newPosts);
  }

  public void appendPosts(List<PostModel> newPosts) {
    List<PostModel> current =
        posts.getValue() != null ? new ArrayList<>(posts.getValue()) : new ArrayList<>();
    current.addAll(newPosts);
    posts.postValue(current);
  }

  public void toggleLike(String postId) {
    // Implement like toggle logic
  }

  public void toggleRepost(String postId) {
    // Implement repost toggle logic
  }

  public void toggleBookmark(String postId) {
    // Implement bookmark toggle logic
  }

  public void reportPost(String postId) {
    // Implement report logic
  }

  public void setError(String error) {
    errors.postValue(error);
  }

  // Getters
  public LiveData<List<PostModel>> getPosts() {
    return posts;
  }

  public LiveData<String> getErrors() {
    return errors;
  }

  public void deletePost(String postId) {
    FirebaseFirestore.getInstance()
        .collection("posts")
        .document(postId)
        .delete()
        .addOnSuccessListener(
            aVoid -> {
              List<PostModel> current = posts.getValue();
              if (current != null) {
                current.removeIf(post -> post.getPostId().equals(postId));
                posts.postValue(current);
              }
            })
        .addOnFailureListener(e -> errors.postValue("Failed to delete post: " + e.getMessage()));
  }

  public void resetState() {
    hasLoaded = false;
    posts.setValue(null);
  }
}

File: app/src/main/java/com/spidroid/starry/ui/home/FeedViewModel.kt
package com.spidroid.starry.ui.home

import com.google.firebase.firestore.FirebaseFirestore

class FeedViewModel : ViewModel() {
    private val posts: MutableLiveData<kotlin.collections.MutableList<PostModel?>?> =
        MutableLiveData<kotlin.collections.MutableList<PostModel?>?>()
    private val errors: MutableLiveData<kotlin.String?> = MutableLiveData<kotlin.String?>()
    private var hasLoaded = false

    fun hasPosts(): kotlin.Boolean {
        return posts.getValue() != null && !posts.getValue().isEmpty()
    }

    fun setPosts(newPosts: kotlin.collections.MutableList<PostModel?>?) {
        hasLoaded = true
        posts.setValue(newPosts)
    }

    fun appendPosts(newPosts: kotlin.collections.MutableList<PostModel?>) {
        val current: kotlin.collections.MutableList<PostModel?> =
            if (posts.getValue() != null) java.util.ArrayList<PostModel?>(posts.getValue()) else java.util.ArrayList<PostModel?>()
        current.addAll(newPosts)
        posts.postValue(current)
    }

    fun toggleLike(postId: kotlin.String?) {
        // Implement like toggle logic
    }

    fun toggleRepost(postId: kotlin.String?) {
        // Implement repost toggle logic
    }

    fun toggleBookmark(postId: kotlin.String?) {
        // Implement bookmark toggle logic
    }

    fun reportPost(postId: kotlin.String?) {
        // Implement report logic
    }

    fun setError(error: kotlin.String?) {
        errors.postValue(error)
    }

    // Getters
    fun getPosts(): LiveData<kotlin.collections.MutableList<PostModel?>?> {
        return posts
    }

    fun getErrors(): LiveData<kotlin.String?> {
        return errors
    }

    fun deletePost(postId: kotlin.String?) {
        FirebaseFirestore.getInstance()
            .collection("posts")
            .document(postId)
            .delete()
            .addOnSuccessListener(
                { aVoid ->
                    val current: kotlin.collections.MutableList<PostModel?>? = posts.getValue()
                    if (current != null) {
                        current.removeIf { post: PostModel? -> post.getPostId() == postId }
                        posts.postValue(current)
                    }
                })
            .addOnFailureListener({ e -> errors.postValue("Failed to delete post: " + e.getMessage()) })
    }

    fun resetState() {
        hasLoaded = false
        posts.setValue(null)
    }
}
